09_advanced-git.qmd

```diff
@@ -0,0 +1,136 @@
+
+### Working with Branches
+
+#### Branching Motivation
+
+The track changes feature in programs like with Microsoft Word follows each and every keystroke or click. In the previous chapter, we compared git to a super-charged version of track changes, and we think this comparison is fair. However, code is different than prose. If a program is not working, it isn’t desirable to document and share a change until that change creates a different, but still working, program. Working with other developers can complicate this desire to always commit sensible changes. 
+
+Meet branching. Branching allows multiple versions of the same program to be developed simultaneously and then merged together when those versions are ready and operational. 
+
+#### Branching Diagrammed 
+Recall the following diagram from last chapter. We need a way to:
+
+1. Create branches
+2. Switch between branches
+3. Combine branches
+
+![A standard git workflow](images/git-branching.png){fig-align="center" width=70%}
+
+
+A second stylized (and cute!) example of this workflow can be seen [in this tweet](https://twitter.com/jay_gee/status/702638177471873024/photo/1) from Jennifer Gilbert. The octopus on the left side of the image represents an existing, operational piece of code. Two programmers create separate copies (branches) of that code, work to create independent features (represented by the heart glasses and a cowboy hat), and then merge those features back to the master branch when those features are operational.
+
+#### How to branch
+`git branch` prints out important information about the available branches in a repo. It is similar to `git status` in that it provides useful information while not making any changes.
+
+`git switch -c <new branch name>` creates a new branch and then navigates you to the new branch.
+
+`git switch <new branch name>` navigates you from your current branch to the specified branch. It is only possible to switch between branches if there are no uncommitted changes. [^checkout]
+
+[^checkout]: `git checkout` is another exceedingly common git command. Many resources on the internet may encourage the usage of `git checkout <branch name>` to switch branches, or `git checkout -b <new branch name>` to create a new branch and then switch to it. This is okay! However, `git checkout` also has other capabilities, and that glut of functionality can be confusing to users. This makes `git switch` the simpler, more modern option.]
+
+
+1. Use `git switch main` to navigate to the main branch. Use `git pull origin main` to update your local version of the `main` branch with any remote changes. 
+2. Use `git switch -c <new branch name>` to create and switch to a new branch with the name `iss<number>`, where <number> is the issue number from GitHub.
+3. Work as if you are on the main branch but push to branch `iss<number>`. 
+
+[Jenny Bryan](https://happygitwithr.com/git-branches.html) provides a more thorough background.
```

Comment:
I've never used `git switch`!!!

Reply:
Yes! I saw Clayton use it semi-recently (maybe in the last year) and did more research which I found enlightening.